message.go

```diff
@@ -4,54 +4,246 @@ import (
 	"errors"
 )
 
-type message struct {
-	method          string
-	seqno           seqNumber
-	res             interface{}
-	err             interface{}
-	remainingFields int
-	decodeSlots     []interface{}
-}
-
-func decodeIntoMessage(dec decoder, m *message) error {
-	for _, s := range m.decodeSlots {
-		if err := decodeMessage(dec, m, s); err != nil {
-			return err
+type RPCMessage interface {
+	Type() MethodType
+	Decode(l int, d decoder, p *protocolHandler, cc *callContainer) error
+	RPCData
+}
+
+type RPCData interface {
+	Name() string
+	Arg() interface{}
+	SeqNo() seqNumber
+	Err() error
+	Res() interface{}
+	ResponseCh() chan RPCMessage
+	MinLength() int
+	DecodeData(int, decoder, *protocolHandler, *callContainer) error
```

Comment:
Also, it's a bit weird clunky to have an interface where you have to call DecodeData before all the other methods work. What about this: repurpose the RPCMessage name to look like:

```
type RPCMessage interface (struct?) {
  DecodeData(...) (RPCData, error)
}
```

and so the caller would do:

```
	if err := d.Decode(&c.typ); err != nil {
		return err
        }

        var msg RPCMessage
	switch c.typ {
	case MethodCall:
		msg = &RPCCallMsg{}
	case MethodResponse:
		msg = &RPCResponseMsg{}
            ...
  	}

        rpcData, err := msg.DecodeData(...)
        // work with rpcData
```

although in that case, you don't even need the RPCMessage struct, really:

```
	if err := d.Decode(&c.typ); err != nil {
		return err
        }

        var rpcData
	switch c.typ {
	case MethodCall:
		rpcData = DecodeCallData(d, ...)
	case MethodResponse:
		rpcData = DecodeResponseData(d, ...)
            ...
  	}
```

Seems easier to follow without all those interfaces in the way.